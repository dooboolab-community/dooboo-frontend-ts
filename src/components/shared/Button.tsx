import React, { Component } from 'react';
import styled from 'styled-components';
import { WhiteButton, TransparentButton } from '../ui/Buttons';

interface IProps {
  id?: string;
  white?: boolean;
  imgSrc?: any;
  text?: string;
  onClick?: () => void;
  isLoading?: boolean;
}

const Text = styled.span`
  font-size: 14px;
  color: rgb(128, 109, 216);
`;

const LogoImg = styled.img`
  position: absolute;
  left: 16px;
  height: 20px;
  width: 20px;
  object-fit: cover
`;

const Spinner = styled.div`
  border: 4px solid #f3f3f3; /* Light grey */
  border-top: 4px solid #3498db; /* Blue */
  border-radius: 50%;
  width: 24px;
  height: 24px;
  animation: spin 1s linear infinite;
  @keyframes spin {
    0% { transform: rotate(0deg); }
    100% { transform: rotate(360deg); }
  }
`;

function Button(props: IProps) {
  const { white, onClick, imgSrc, text } = props;
  if (props.white) {
    return (
      <WhiteButton
        style={{ height: '60px' }}
        onClick={onClick}
      >
        {
          props.isLoading
            ? <Spinner id='spinner'/>
            : <div>
              {
                props.imgSrc
                  ? <LogoImg
                    src={props.imgSrc}
                  />
                  : null
              }
              <Text>{props.text}</Text>
            </div>
        }
      </WhiteButton>
    );
  }
  return (
    <TransparentButton
      style={{ height: '60px' }}
      onClick={onClick}
    >
      {
        props.isLoading
          ? <Spinner id='spinner'/>
          : <div>
            {
              props.imgSrc
                ? <LogoImg
                  src={props.imgSrc}
                />
                : null
            }
            <Text style={{ color: 'white' }}>{props.text}</Text>
          </div>
      }
    </TransparentButton>
  );
}

export default Button;
