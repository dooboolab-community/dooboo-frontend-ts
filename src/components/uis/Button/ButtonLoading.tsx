import type {ReactElement} from 'react';
import styled from '@emotion/styled';

const Spinner = styled.div`
  align-self: center;
  border: 4px solid #f3f3f3; /* Light grey */
  border-top: 4px solid #3498db; /* Blue */
  border-radius: 50%;
  width: 24px;
  height: 24px;
  animation: spin 1s linear infinite;
  @keyframes spin {
    0% {
      transform: rotate(0deg);
    }
    100% {
      transform: rotate(360deg);
    }
  }
`;

const ButtonLoading = (): ReactElement => {
  return <Spinner id="spinner" />;
};

export default ButtonLoading;
