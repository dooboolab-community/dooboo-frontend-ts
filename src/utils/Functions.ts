const getStorage = (key: string): string | null => {
  return localStorage.getItem(key);
};

const setStorage = (key: string, value: string): void => {
  return localStorage.setItem(key, value);
};

const destroyStorage = (key: string): void => {
  return localStorage.removeItem(key);
};

const getSessionStorage = (key: string): string | null => {
  return sessionStorage.getItem(key);
};

const setSessionStorage = (key: string, value: string): void => {
  return sessionStorage.setItem(key, value);
};

const destroySessionStorage = (key: string): void => {
  return sessionStorage.removeItem(key);
};

const checkImageExists = (
  url: string,
  callback: (err: Error | null, val: boolean) => void,
): void => {
  const img = new Image();

  img.onload = (): void => {
    callback(null, true);
  };

  img.onerror = (): void => {
    callback(new Error('error'), false);
  };

  img.src = url;
};

export {
  getStorage,
  setStorage,
  destroyStorage,
  getSessionStorage,
  setSessionStorage,
  destroySessionStorage,
  checkImageExists,
};
