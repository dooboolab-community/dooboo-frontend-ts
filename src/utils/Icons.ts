import icFacebookW from '../../assets/icons/ic-facebook-white.png';
import icFacebookW2x from '../../assets/icons/ic-facebook-white@2x.png';
import icFacebookW3x from '../../assets/icons/ic-facebook-white@3x.png';
import icGoogleW from '../../assets/icons/ic-google-white.png';
import icGoogleW2x from '../../assets/icons/ic-google-white@2x.png';
import icGoogleW3x from '../../assets/icons/ic-google-white@3x.png';

const isRetina = (first: string, second?: string, third?: string): string => {
  if (window.devicePixelRatio >= 1 && third) {
    return third;
  }
  if (window.devicePixelRatio >= 0.5 && second) {
    return second;
  }
  return first;
};

export const IC_FACEBOOK_W = isRetina(
  icFacebookW,
  icFacebookW2x,
  icFacebookW3x,
);

export const IC_GOOGLE_W = isRetina(icGoogleW, icGoogleW2x, icGoogleW3x);
