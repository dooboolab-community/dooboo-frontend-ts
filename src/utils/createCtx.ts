import {
  type ProviderExoticComponent,
  type ProviderProps,
  createContext,
  useContext,
} from 'react';

type CreateCtx<A> = readonly [
  () => A,
  ProviderExoticComponent<ProviderProps<A | undefined>>,
];

// create context with no upfront defaultValue
// without having to do undefined check all the time
function createCtx<A>(): CreateCtx<A> {
  const ctx = createContext<A | undefined>(undefined);

  function useCtx(): A {
    const c = useContext(ctx);

    if (!c) {
      throw new Error('useCtx must be inside a Provider with a value');
    }

    return c;
  }

  // make TypeScript infer a tuple, not an array of union types
  return [useCtx, ctx.Provider] as const;
}

export default createCtx;
