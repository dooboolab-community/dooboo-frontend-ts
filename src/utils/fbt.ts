import {init} from 'fbt';
import intl from '../../assets/translatedFbts.json';

const DEFAULT_LOCALE = 'en';

export const viewerContext = {
  locale: DEFAULT_LOCALE,
};

export const initFbt = (): void => {
  if (navigator) {
    viewerContext.locale = navigator.language.substr(0, 2);
  }

  init({
    translations: intl,
    hooks: {
      // @ts-ignore
      getViewerContext: (): {locale: string} => viewerContext,
    },
  });
};

export const LOCALES = Object.freeze({
  ko: Object.freeze({
    bcp47: 'ko',
    rtl: false,
  }),
  en: Object.freeze({
    bcp47: 'en',
    rtl: false,
  }),
});

export enum Locale {
  KO = 'ko',
  EN = 'en',
}

export const changeLocale = (locale: Locale): void => {
  viewerContext.locale = locale;

  const html = document.getElementsByTagName('html')[0];

  if (html != null) {
    html.lang = LOCALES[locale].bcp47;
  }

  document.body.className = LOCALES[locale].rtl ? 'rtl' : 'ltr';
};
