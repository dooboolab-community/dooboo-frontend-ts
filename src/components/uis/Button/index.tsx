import {CSSProperties, ReactElement} from 'react';

import ButtonLoading from './ButtonLoading';
import {ButtonPrimary} from '../Styles';
import styled from '@emotion/styled';

interface Props {
  testID?: string;
  id?: string;
  style?: CSSProperties;
  imgSrc?: string;
  text?: string | ReactElement;
  onClick?: () => void;
  isLoading?: boolean;
}

const ButtonWrapper = styled.div`
  flex: 1;
  height: 100%;

  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
`;

const PrimaryText = styled.span`
  font-size: 14px;
  color: ${({theme}) => theme.btnPrimaryFont};
  margin: auto;
`;

const LogoImg = styled.img`
  position: absolute;
  left: 20px;
  height: 20px;
  width: 20px;
  object-fit: cover;
`;

function Button(props: Props): ReactElement {
  const {testID, onClick, imgSrc, text, style, isLoading} = props;

  return (
    <ButtonPrimary data-testid={testID} style={style} onClick={onClick}>
      {isLoading ? (
        <ButtonLoading />
      ) : (
        <ButtonWrapper>
          {imgSrc ? <LogoImg src={imgSrc} /> : null}
          <PrimaryText>{text}</PrimaryText>
        </ButtonWrapper>
      )}
    </ButtonPrimary>
  );
}

Button.defaultProps = {
  style: {display: 'flex', height: '60px'},
};

export default Button;
