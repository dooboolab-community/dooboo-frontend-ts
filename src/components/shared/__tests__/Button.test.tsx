import React from 'react';
import Button from '../Button';

import * as renderer from 'react-test-renderer';
import { IC_FACEBOOK_W } from '../../../utils/Icons';

// test for the pure component
describe('[Button] shared component test', () => {
  let tree: any;
  const component: any = <Button white={true} text='Button 2nd test' />;

  it('component and snapshot matches', () => {
    tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('[Transparent] interaction', () => {
  let count = 1;
  const onClick = () => {
    count++;
  };

  let rendered: renderer.ReactTestRenderer;
  let root: renderer.ReactTestInstance;
  const component: any = <Button onClick={onClick}/>;

  beforeAll(() => {
    rendered = renderer.create(component);
    root = rendered.root;
  });

  it('Simulate onClick', () => {
    const button = root.find(
      (el: any) => el.type === 'button',
    );
    button.props.onClick();
    expect(count).toBe(2);
  });
});

describe('[WhiteButton] interaction', () => {
  let count = 1;
  const onClick = () => {
    count++;
  };

  let rendered: renderer.ReactTestRenderer;
  let root: renderer.ReactTestInstance;
  let component: any;

  it('should simulate onClick', () => {
    component = <Button white={true} onClick={onClick}/>;
    rendered = renderer.create(component);
    root = rendered.root;

    const button = root.find(
      (el: any) => el.type === 'button',
    );
    button.props.onClick();
    expect(count).toBe(2);
  });

  it('should render isLoading status', () => {
    component = <Button white={true} isLoading={true}/>;
    rendered = renderer.create(component);
    root = rendered.root;
  });

  it('should render img status', () => {
    component = <Button white={true} imgSrc={IC_FACEBOOK_W}/>;
    rendered = renderer.create(component);
    root = rendered.root;
  });
});
