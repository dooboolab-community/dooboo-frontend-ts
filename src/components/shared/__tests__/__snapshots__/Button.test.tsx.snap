// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Button] shared component test component and snapshot matches 1`] = `
<button
  className="sc-bwzfXH jGtnoc"
  style={
    Object {
      "height": "60px",
    }
  }
>
  <div>
    <span
      className="sc-htpNat gUfFlb"
    >
      Button 2nd test
    </span>
  </div>
</button>
`;
